Better tests for utils

acforvs opened the pull request:
Hi, hopefully this one closes #404

hbq1 commented on optax/_src/utils_test.py:
```diff
@@ -61,5 +62,70 @@ def fn(inputs):
+
+class MultiNormalDiagFromLogScaleTest(parameterized.TestCase):
+
+  @staticmethod
```
No need to use a @staticmethod here (see https://google.github.io/styleguide/pyguide.html#217-function-and-method-decorators), consider making it a class function.

hbq1 commented on optax/_src/utils_test.py:
```diff
@@ -61,5 +62,70 @@ def fn(inputs):
+    dist = utils.multi_normal(loc, scale)
+    self.assertIsInstance(dist, utils.MultiNormalDiagFromLogScale)
+    mean, log_scale = dist.params
+    self.assertTrue(tuple(mean.shape) == tuple_shape(loc_shape))
```
Consider using `self.assertEqual(a, b)` instead of `self.assertTrue(a == b)` for better error messages.

hbq1 commented on optax/_src/utils_test.py:
```diff
@@ -61,5 +62,70 @@ def fn(inputs):
+  )
+  def test_init_unsuccessful_broadcast(self, loc_shape, scale_shape):
+    loc, scale = self._get_loc_scale(loc_shape, scale_shape)
+    with self.assertRaises(ValueError):
```
`self.assertRaisesRegex(ValueError, <regexp>)` would be more safe here

hbq1 commented on optax/_src/utils_test.py:
```diff
@@ -61,5 +62,70 @@ def fn(inputs):
+    loc_shape = scale_shape = (2, 3)
+    loc, scale = self._get_loc_scale(loc_shape, scale_shape)
+    dist = utils.multi_normal(loc, scale)
+    samples = dist.sample(sample_shape, jax.random.PRNGKey(239))
```
SPB Lyceum 239? :)

acforvs replied:
yep :)

hbq1 replied:
very cool :)

hbq1 commented on optax/_src/utils_test.py:
```diff
@@ -61,5 +62,70 @@ def fn(inputs):
+    loc, scale = self._get_loc_scale(loc_shape, scale_shape)
+    dist = utils.multi_normal(loc, scale)
+    samples = dist.sample(sample_shape, jax.random.PRNGKey(239))
+    self.assertTrue(samples.shape == tuple(sample_shape) + loc_shape)
```
ditto re: `assertEqual`

hbq1 commented on optax/_src/utils_test.py:
```diff
@@ -61,5 +62,70 @@ def fn(inputs):
+    probs = dist.log_prob(jnp.ones(shape=shape, dtype=jnp.float32))
+
+    self.assertIsInstance(probs, jnp.DeviceArray)
+    self.assertFalse(bool(probs.shape))
```
ditto

hbq1 replied:
Yes, thanks!

hbq1 commented:
Hi @acforvs, many thanks for your PR! 
It looks good to me; I left only a few comments. Could you also add tests for the rest of helpers from utils.py, so we could close the original issue?

acforvs commented:
Hi @hbq1 , thank you for the review!

Sure thing, I'll add the tests for the remaining functions, but I have some questions about them:
1. The docstring for `set_diags` [states](https://github.com/deepmind/optax/blob/master/optax/_src/utils.py#L56) that the shape of `new_diags` should be `NxD`. However, `new_diags` is only used in one place where it is [flattened](https://github.com/deepmind/optax/blob/master/optax/_src/utils.py#L70).

As a result, examples like
```python
a = jnp.ones(shape=(3, 2, 2))
diags = jnp.array([[[3, 3, 3, 3, 3, 3]]])
print(a.shape, diags.shape)  # prints (3, 2, 2) (1, 1, 6)
optax._src.utils.set_diags(a, diags)
```
or 
```python
a = jnp.ones(shape=(3, 2, 2))
diags = jnp.array([[3, 3, 3], [3, 3, 3]])
print(a.shape, diags.shape)  # prints (3, 2, 2) (2, 3)
optax._src.utils.set_diags(a, diags)
```
work just fine

**Is this by design or should we raise an error in case of a shape mismatch?** 
Currently, the function is used internally in two places: [1](https://github.com/deepmind/optax/blob/master/optax/_src/stochastic_gradient_estimators.py#L237), [2](https://github.com/deepmind/optax/blob/master/optax/_src/stochastic_gradient_estimators.py#L302). The comments specify that the dimension of the passed variable is indeed `NxD` there, so I'd suggest adding another `assert`. What are your thoughts?

2. `cast_tree` uses `.astype` which wouldn't work for simple datatypes, such as 
```python
tree = dict(a=2.5, b=dict(c=-2.5))
tree = jax.tree_util.tree_map(lambda x : x, tree)
optax._src.utils.cast_tree(tree, int)
```
Would it be okay to only test this function for dicts of jnp.arrays of differerent dtypes? Something like 
```python
tree = dict(a=jnp.array(2.5), b=dict(c=jnp.array(-2.5)))
tree = jax.tree_util.tree_map(lambda x : x, tree)
optax._src.utils.cast_tree(tree, int)
```

hbq1 commented:
Great, thank you!

1. It is used in some internal code, so the correct behaviour would be to raise an exception in case of shape mismatch; it'd be great if you add this to the implementation!
2. Testing only with trees of `jnp.array`s would suffice.

acforvs commented:
I've fixed the errors and added tests for the rest of the utils.py functions. 
I also changed `assert` to a `ValueError` in `set_diags`. Please let me know how it looks

hbq1 reviewed (approved):
Thanks for the changes and for the nice PR 👍
